openmls/src/group/core_group/mod.rs

```diff
@@ -1006,12 +996,10 @@ impl CoreGroup {
             };
             // Sign to-be-signed group info.
             Some(group_info_tbs.sign(signer)?)
-        } else {
-            None
         };
 
         // Check if new members were added and, if so, create welcome messages
-        let welcome_option = if !encrypted_secrets.is_empty() {
```

Comment:
Is there a reason you moved the condition to the end? It seems that if the condition is false, we're doing quite a few computations only to throw them away afterwards. Although it's of course possible that this is optimized away by LLVM.

Reply:
Good catch! I thought there was a reason but `diff.encrypt_group_secrets` does way less than I thought. By replacing `encrypted_group_secrets.is_empty()` with `apply_proposals_values.invitation_list.is_empty()` we can spare us the work. Regarding the joiner_secret: I discussed that with Franziskus and we don't think it is possible to get this working without cloning as the `joiner_secret` is effectively required two times: once to derive the welcome_secret and once to derive the secret used to encrypt the group secrets (label "Welcome" vs "welcome"). I made `KeySchedule::init` take a reference now to spare us the clone. But I think we would need to refactor the key schedule to make this nicer. But I would propose to do it in a follow-up. What do you think?